File: mobile/lib/models/family.dart
import 'user.dart';

class FamilyMember {
  final String id;
  final String name;
  final String phone;
  final String relationship;
  final DateTime joinedAt;
  final FamilyPermissions permissions;

  const FamilyMember({
    required this.id,
    required this.name,
    required this.phone,
    required this.relationship,
    required this.joinedAt,
    required this.permissions,
  });

  factory FamilyMember.fromJson(Map<String, dynamic> json) {
    return FamilyMember(
      id: json['id'] as String,
      name: json['name'] as String,
      phone: json['phone'] as String,
      relationship: json['relationship'] as String,
      joinedAt: DateTime.parse(json['joined_at'] as String),
      permissions: FamilyPermissions.fromJson(
        json['permissions'] as Map<String, dynamic>,
      ),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'phone': phone,
    'relationship': relationship,
    'joined_at': joinedAt.toIso8601String(),
    'permissions': permissions.toJson(),
  };
}

class FamilyPermissions {
  final bool viewProfile;
  final bool viewPhotos;
  final bool viewMatches;
  final bool suggestProfiles;
  final bool viewActivity;
  final bool viewScores;
  final bool viewChat;
  final bool receiveNotifications;

  const FamilyPermissions({
    this.viewProfile = true,
    this.viewPhotos = true,
    this.viewMatches = false,
    this.suggestProfiles = true,
    this.viewActivity = false,
    this.viewScores = true,
    this.viewChat = false,
    this.receiveNotifications = false,
  });

  factory FamilyPermissions.fromJson(Map<String, dynamic> json) {
    return FamilyPermissions(
      viewProfile: json['view_profile'] as bool? ?? true,
      viewPhotos: json['view_photos'] as bool? ?? true,
      viewMatches: json['view_matches'] as bool? ?? false,
      suggestProfiles: json['suggest_profiles'] as bool? ?? true,
      viewActivity: json['view_activity'] as bool? ?? false,
      viewScores: json['view_scores'] as bool? ?? true,
      viewChat: json['view_chat'] as bool? ?? false,
      receiveNotifications: json['receive_notifications'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() => {
    'view_profile': viewProfile,
    'view_photos': viewPhotos,
    'view_matches': viewMatches,
    'suggest_profiles': suggestProfiles,
    'view_activity': viewActivity,
    'view_scores': viewScores,
    'view_chat': viewChat,
    'receive_notifications': receiveNotifications,
  };

  FamilyPermissions copyWith({
    bool? viewProfile,
    bool? viewPhotos,
    bool? viewMatches,
    bool? suggestProfiles,
    bool? viewActivity,
    bool? viewScores,
    bool? viewChat,
    bool? receiveNotifications,
  }) {
    return FamilyPermissions(
      viewProfile: viewProfile ?? this.viewProfile,
      viewPhotos: viewPhotos ?? this.viewPhotos,
      viewMatches: viewMatches ?? this.viewMatches,
      suggestProfiles: suggestProfiles ?? this.suggestProfiles,
      viewActivity: viewActivity ?? this.viewActivity,
      viewScores: viewScores ?? this.viewScores,
      viewChat: viewChat ?? this.viewChat,
      receiveNotifications: receiveNotifications ?? this.receiveNotifications,
    );
  }
}

class FamilySuggestion {
  final String id;
  final FamilyMember suggestedBy;
  final User suggestedUser;
  final String? note;
  final DateTime suggestedAt;

  const FamilySuggestion({
    required this.id,
    required this.suggestedBy,
    required this.suggestedUser,
    this.note,
    required this.suggestedAt,
  });

  factory FamilySuggestion.fromJson(Map<String, dynamic> json) {
    return FamilySuggestion(
      id: json['id'] as String,
      suggestedBy: FamilyMember.fromJson(
        json['suggested_by'] as Map<String, dynamic>,
      ),
      suggestedUser: User.fromJson(
        json['suggested_user'] as Map<String, dynamic>,
      ),
      note: json['note'] as String?,
      suggestedAt: DateTime.parse(json['suggested_at'] as String),
    );
  }
}

class FamilyInvite {
  final String code;
  final DateTime expiresAt;

  const FamilyInvite({required this.code, required this.expiresAt});

  factory FamilyInvite.fromJson(Map<String, dynamic> json) {
    return FamilyInvite(
      code: json['code'] as String,
      expiresAt: DateTime.parse(json['expires_at'] as String),
    );
  }
}

File: ios/MitiMaiti/Models/Family.swift
import Foundation

struct FamilyPermissions: Codable, Hashable {
    var canViewProfile: Bool
    var canViewPhotos: Bool
    var canViewBasics: Bool
    var canViewSindhi: Bool
    var canViewMatches: Bool
    var canSuggest: Bool
    var canViewCulturalScore: Bool
    var canViewKundli: Bool

    static let allEnabled = FamilyPermissions(
        canViewProfile: true, canViewPhotos: true, canViewBasics: true,
        canViewSindhi: true, canViewMatches: true, canSuggest: true,
        canViewCulturalScore: true, canViewKundli: true
    )

    static let allDisabled = FamilyPermissions(
        canViewProfile: false, canViewPhotos: false, canViewBasics: false,
        canViewSindhi: false, canViewMatches: false, canSuggest: false,
        canViewCulturalScore: false, canViewKundli: false
    )

    var enabledCount: Int {
        [canViewProfile, canViewPhotos, canViewBasics, canViewSindhi,
         canViewMatches, canSuggest, canViewCulturalScore, canViewKundli]
            .filter { $0 }.count
    }

    /// Maps each permission key path to its (icon, title, description).
    static let permissionDescriptions: [(keyPath: WritableKeyPath<FamilyPermissions, Bool>, icon: String, title: String, description: String)] = [
        (\.canViewProfile, "eye", "View Profile", "Can see your full profile details"),
        (\.canViewPhotos, "camera", "View Photos", "Can browse your uploaded photos"),
        (\.canViewBasics, "briefcase", "View Basics", "Can see education, occupation & lifestyle"),
        (\.canViewSindhi, "moon.stars", "Sindhi Identity", "Can view your Sindhi heritage details"),
        (\.canViewMatches, "heart", "View Matches", "Can see your current matches"),
        (\.canSuggest, "message", "Suggest Profiles", "Can recommend profiles for you"),
        (\.canViewCulturalScore, "star", "Cultural Score", "Can view your cultural compatibility score"),
        (\.canViewKundli, "chart.bar", "Kundli Score", "Can view your kundli compatibility score"),
    ]
}

struct FamilyMember: Identifiable, Codable, Hashable {
    let id: String
    let name: String
    let phone: String
    var relationship: String
    var status: FamilyMemberStatus
    var permissions: FamilyPermissions
    let joinedAt: Date

    init(
        id: String = UUID().uuidString,
        name: String,
        phone: String,
        relationship: String,
        status: FamilyMemberStatus = .active,
        permissions: FamilyPermissions = .allEnabled,
        joinedAt: Date = Date()
    ) {
        self.id = id
        self.name = name
        self.phone = phone
        self.relationship = relationship
        self.status = status
        self.permissions = permissions
        self.joinedAt = joinedAt
    }
}

struct FamilyInvite: Codable, Hashable {
    let code: String
    let deepLink: String
    let expiresAt: Date
    let currentMembers: Int
    let maxMembers: Int

    init(
        code: String = "MM-\(String(Int.random(in: 100000...999999)))",
        deepLink: String = "",
        expiresAt: Date = Calendar.current.date(byAdding: .hour, value: 48, to: Date())!,
        currentMembers: Int = 1,
        maxMembers: Int = 3
    ) {
        self.code = code
        self.deepLink = "mitimaiti://family/join?code=\(code)"
        self.expiresAt = expiresAt
        self.currentMembers = currentMembers
        self.maxMembers = maxMembers
    }
}

struct FamilySuggestion: Identifiable, Codable, Hashable {
    let id: String
    let suggestedBy: FamilyMember
    let suggestedUser: User
    let note: String?
    let suggestedAt: Date

    init(
        id: String = UUID().uuidString,
        suggestedBy: FamilyMember,
        suggestedUser: User,
        note: String? = nil,
        suggestedAt: Date = Date()
    ) {
        self.id = id
        self.suggestedBy = suggestedBy
        self.suggestedUser = suggestedUser
        self.note = note
        self.suggestedAt = suggestedAt
    }
}
